### PR TITLE
rbm python version

### PR DESCRIPTION
**Pull request by lzjpaul**
python version of rbm

this code is referring to hinton's science paper
http://www.cs.toronto.edu/~hinton/science.pdf

### CODE REVIEW

File: examples/rbm-python/train.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    hdim = 1000
+    vdim = 784
+
+    shape = (vdim, hdim)
+    tweight = tensor.Tensor(shape)
+    initializer.gaussian(tweight, float(0.0), float(0.1)) # ?? is this correct? should have a right hand side?
+    gweight = np.zeros([vdim, hdim])
+    gweight = gweight.astype(np.float32) # ?? initialization should be correct
+    tgweight = tensor.Tensor(gweight.shape)
+    tgweight.copy_from_numpy(gweight)
```
Comment (nudles) [inline examples/rbm-python/train.py]:
```
gweight = np.zeros([], dtype=np.float32)
tgweight = tensor.from_numpy(gweight)
```

File: examples/rbm-python/train.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        for b in range(num_train_batch):
+            # positive phase
+            data = train_x[:, :, b : (b + 1)]
+            data = data.reshape(data.shape[0], data.shape[1])
+#            print "data shape = \n", data.shape
+            tdata = tensor.Tensor(data.shape)
+            tdata.copy_from_numpy(data)
+#            print "tweight shape: \n", tweight.shape
+            tposhidprob = tensor.mult(tdata, tweight)
+            poshidprob = tensor.to_numpy(tposhidprob) + np.matlib.repmat(tensor.to_numpy(thbias), batch_size, 1) # ?? repmat shape correct??
```
Comment (nudles) [inline examples/rbm-python/train.py]: why converting tensor to numpy array?
there is tensor::add_row().

File: examples/rbm-python/train.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            tgvbias_sample = (tensor.sum(tdata, 0) - tensor.sum(tnegdata, 0)) # sum_rows is it row??
+            tgvbias_momentum *= momentum
+            tgvbias_sample *= float(bias_lr/float(batch_size)) # ?? float conversion??
+            tgvbias = tgvbias_momentum + tgvbias_sample
+            # tgvbias = momentum * tgvbias + (bias_lr/float(batch_size)) * (sum_rows(tdata) - sum_rows(tnegdata))
+            tghbias_momentum = tghbias
+            tghbias_sample = (tensor.sum(tposhidprob, 0) - tensor.sum(tneghidprob, 0)) # sum_rows is it row??
+            tghbias_momentum *= momentum
+            tghbias_sample *= float(bias_lr/float(batch_size)) # ?? float conversion avoiding all 0s??
+            tghbias = tghbias_momentum + tghbias_sample
```
Comment (nudles) [inline examples/rbm-python/train.py]: you can use sgd in optimizer.py to update parameters...

File: examples/rbm-python/train.py (hunk 2 above)
Comment (lzjpaul) [inline examples/rbm-python/train.py]: yes, sure, will use add_row() instead

File: examples/rbm-python/train.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    weight_lr = 0.1;   # Learning rate for weights 
+    bias_lr = 0.1;   # Learning rate for biases of visible units 
+    weight_decay  = 0.0002;   
+    init_momentum  = 0.5;
+    final_momentum  = 0.9;
+    hdim = 1000
+    vdim = 784
+
+    shape = (vdim, hdim)
+    tweight = tensor.Tensor(shape)
```
Comment (nudles) [inline examples/rbm-python/train.py]: tweight = tensor.Tensor((vidm, hdim))

File: examples/rbm-python/train.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    initializer.gaussian(tweight, float(0.0), float(0.1)) # ?? is this correct? should have a right hand side?
+    gweight = np.zeros([vdim, hdim])
+    gweight = gweight.astype(np.float32) # ?? initialization should be correct
```
Comment (nudles) [inline examples/rbm-python/train.py]: gweight = np.zeros((vdim, hdim), dtype=np.float32)

File: examples/rbm-python/train_to_print.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+#    hbias = np.zeros(hdim)
+#    hbias = hbias.astype(np.float32)
+#    thbias = tensor.Tensor(hbias.shape)
+#    thbias.copy_from_numpy(hbias) # ?? is this copy data from?? what does offset mean?
+#    ghbias = np.zeros(hdim)
+#    ghbias = ghbias.astype(np.float32)
+#    tghbias = tensor.Tensor(ghbias.shape)
+#    tghbias.copy_from_numpy(ghbias) # ?? is this copy data from?? what does offset mean?
+    gweight = np.zeros([vdim, hdim], dtype = np.float32) # ?? initialization correct??
+    tgweight = tensor.from_numpy(gweight)
```
Comment (nudles) [inline examples/rbm-python/train_to_print.py]: you can also merge the above two lines into a single line,

gweight = tensor.from_numpy(np.zeros((vdim, hdim), dtype=np.float32))

File: examples/rbm-python/train_to_print.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+from singa import initializer
+from singa import utils
+from singa import optimizer
+from singa import device
+from singa import tensor
+from singa.proto import core_pb2
+
+
+
+# ??? need normalization??
```
Comment (nudles) [inline examples/rbm-python/train_to_print.py]: Does the matlab code have data normalization?

File: examples/rbm-python/train_to_print.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+            tposhidsample.copy_from_numpy(poshidsample)
+            # negative phase
+            tnegdata = tensor.mult(tposhidsample, tweight.transpose())
+            # negdata = tensor.to_numpy(tnegdata) + numpy.matlib.repmat(tensor.to_numpy(tvbias), batch_size, 1)
+            # tnegdata.copy_from_numpy(negdata)
+            tnegdata.add_row(tvbias)
+            tnegdata = tensor.sigmoid(tnegdata)
+
+            tneghidprob = tensor.mult(tnegdata, tweight)
+            neghidprob = tensor.to_numpy(tneghidprob) + numpy.matlib.repmat(tensor.to_numpy(thbias), batch_size, 1) # ?? repmat shape correct??
```
Comment (nudles) [inline examples/rbm-python/train_to_print.py]: why use numpy.matlib.repmat?
how about add_row or add_column

File: examples/rbm-python/train_to_print.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+            # print "error: ", error
+            # print "errorsum: \n", errorsum
+
+            if epoch > 5:
+                momentum = final_momentum
+            else:
+                momentum = init_momentum
+            # tgweight_momentum = tgweight
+            # tgweight_decay = tgweight
+            tgweight_momentum = tensor.Tensor(tgweight.shape)
```
Comment (nudles) [inline examples/rbm-python/train_to_print.py]: why not use optimizer.sgd()

File: examples/rbm-python/train_to_print_SGD_pythonmnist.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    traindata = train_set[0].astype(np.float32)
+    validdata = valid_set[0].astype(np.float32)
+    trainvaliddata = np.concatenate((traindata, validdata), axis=0)
+    return trainvaliddata
+
+
+
+def train(data_dir, num_epoch=10, batch_size=100):
+    print 'Start intialization............'
+    weight_lr = 0.1   # Learning rate for weights
```
Comment (nudles) [inline examples/rbm-python/train_to_print_SGD_pythonmnist.py]: since these two values are the same, then no need to use two variables.

File: examples/rbm-python/train_to_print_SGD_pythonmnist.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+                print "batch: \n", b
+
+
+            tdata = tensor.from_numpy(train_x[ (b * batch_size): ((b + 1) * batch_size), : ])
+            tposhidprob = tensor.mult(tdata, tweight)
+            tposhidprob.add_row(thbias)
+            tposhidprob = tensor.sigmoid(tposhidprob)
+            tposhidrandom = tensor.Tensor(tposhidprob.shape)
+            initializer.uniform(tposhidrandom,0.0, 1.0)
+            poshidsample = tensor.to_numpy(tposhidprob) > tensor.to_numpy(tposhidrandom)
```
Comment (nudles) [inline examples/rbm-python/train_to_print_SGD_pythonmnist.py]: I think you can compare tensors directly.
no need to convert them into numpy.

File: examples/rbm-python/train_to_print_SGD_pythonmnist.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+            tnegdata = tensor.sigmoid(tnegdata)
+
+            tneghidprob = tensor.mult(tnegdata, tweight)
+            tneghidprob.add_row(thbias) 
+            tneghidprob = tensor.sigmoid(tneghidprob)
+            error = tensor.sum(tensor.eltwise_mult((tdata - tnegdata),(tdata - tnegdata)))
+            errorsum = error + errorsum
+           
+            tgweight = tensor.mult(tnegdata.transpose(), tneghidprob) - tensor.mult(tdata.transpose(), tposhidprob)
+            tgweight *= float(1.0/float(batch_size))
```
Comment (nudles) [inline examples/rbm-python/train_to_print_SGD_pythonmnist.py]: 100 and 106 can be merged into a single line
```
opt.apply_with_lr(epoch, weight_lr / batch_size, tgweight, tweight, 'weight')
```

File: examples/rbm-python/train_to_print_SGD_pythonmnist.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+            
+            tgvbias = tensor.sum(tnegdata, 0) - tensor.sum(tdata, 0)
+            tgvbias *= float(1.0/float(batch_size))
+            tghbias = tensor.sum(tneghidprob, 0) - tensor.sum(tposhidprob, 0)
+            tghbias *= float(1.0/float(batch_size))
+            opt.apply_with_lr(epoch, weight_lr, tgweight, tweight, '')
```
Comment (nudles) [inline examples/rbm-python/train_to_print_SGD_pythonmnist.py]: Will the learning rate decay?
Or is it the same throughout the training?
